src/ring_theory/dedekind_domain/factorization.lean

```diff
@@ -0,0 +1,196 @@
+maximal ideals of `R`, where the exponents `n_v` are natural numbers.
+
+TODO: Extend the results in this file to fractional ideals of `R`.
+
+## Main results
+- `ideal.finite_factors` : Only finitely many maximal ideals of `R` divide a given nonzero ideal.
+- `ideal.factorization` : The ideal `I` equals the finprod `∏_v v^(val_v(I))`, where `val_v(I)`
+  denotes the multiplicity of `v` in the factorization of `I` and `v` runs over the maximal
+  ideals of `R`.
+
+## Tags
+dedekind domain, ideal, factorization
+-/
+
+noncomputable theory
+open_locale big_operators classical
+
+open set function unique_factorization_monoid
+
+/-! ### Factorization of ideals of Dedekind domains -/
+
+open is_dedekind_domain
+
+variables {R : Type*} [comm_ring R] [is_domain R] [is_dedekind_domain R] {K : Type*} [field K]
+  [algebra R K] [is_fraction_ring R K] (v : height_one_spectrum R)
+
+/-- Only finitely many maximal ideals of `R` divide a given nonzero ideal. -/
+lemma ideal.finite_factors {I : ideal R} (hI : I ≠ 0) :
+  { v : height_one_spectrum R | v.as_ideal ∣ I }.finite :=
+begin
+  haveI h_fin := fintype_subtype_dvd I hI,
+  let f' : finset (ideal R) := finset.map
+    ⟨(λ J : {x // x ∣ I}, J.val), subtype.coe_injective⟩ h_fin.elems,
+  have h_eq : { v : height_one_spectrum R | v.as_ideal ∣ I } =
+    { v : height_one_spectrum R | v.as_ideal ∈ f' },
+  { ext v,
+    simp_rw [finset.mem_map, exists_prop, subtype.exists, embedding.coe_fn_mk,
+      exists_and_distrib_right, exists_eq_right],
+    exact ⟨λ h, ⟨h, fintype.complete ⟨v.as_ideal, h⟩⟩, λ ⟨hv, h⟩, hv⟩, },
+  rw h_eq,
+  have hv_inj : injective (λ (v : height_one_spectrum R), v.as_ideal),
+  { intros v w hvw,
+    exact (height_one_spectrum.ext_iff v w).mpr hvw, },
+  exact finite.preimage_embedding ⟨(λ v : height_one_spectrum R, v.as_ideal), hv_inj⟩
+    (finite_mem_finset (f')),
+end
+
+/-- For every nonzero ideal `I` of `v`, there are finitely many maximal ideals `v` such that the
+  multiplicity of `v` in the factorization of `I`, denoted `val_v(I)`, is nonzero. -/
+lemma associates.finite_factors {I : ideal R} (hI : I ≠ 0) :
+  ∀ᶠ (v : height_one_spectrum R) in filter.cofinite,
+    ((associates.mk v.as_ideal).count (associates.mk I).factors : ℤ) = 0 :=
+begin
+  have h_supp : {v : height_one_spectrum R |
+    ¬((associates.mk v.as_ideal).count (associates.mk I).factors : ℤ) = 0} =
+    { v : height_one_spectrum R | v.as_ideal ∣ I },
+  { ext v,
+    simp_rw int.coe_nat_eq_zero,
+    exact associates.count_ne_zero_iff_dvd hI v.irreducible, },
+  rw [filter.eventually_cofinite, h_supp],
+  exact ideal.finite_factors hI,
+end
+
+namespace ideal
+
+/-- For every nonzero ideal `I` of `v`, there are finitely many maximal ideals `v` such that
+  `v^(val_v(I))` is not the unit ideal. -/
+lemma finite_mul_support {I : ideal R} (hI : I ≠ 0) :
+  (mul_support (λ (v : height_one_spectrum R),
+    v.as_ideal^(associates.mk v.as_ideal).count (associates.mk I).factors)).finite :=
```

Comment:
I'd keep spaces around the `^` since otherwise suggests the brackets are placed as `(v.as_ideal ^ (associates.mk v.as_ideal).count) (associates.mk I).factors`.
```suggestion
    v.as_ideal ^ (associates.mk v.as_ideal).count (associates.mk I).factors)).finite :=
```